src/sqlparser/src/parser_v2/expr.rs

```diff
@@ -0,0 +1,60 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use winnow::combinator::{cut_err, opt, preceded, repeat, trace};
+use winnow::{PResult, Parser};
+
+use super::TokenStream;
+use crate::ast::Expr;
+use crate::keywords::Keyword;
+
+fn expr<S>(input: &mut S) -> PResult<Expr>
+where
+    S: TokenStream,
+{
+    // TODO: implement this function using combinator style.
+    trace("expr", |input: &mut S| {
+        input.parse_v1(|parser| parser.parse_expr())
+    })
+    .parse_next(input)
+}
+
+pub fn expr_case<S>(input: &mut S) -> PResult<Expr>
+where
+    S: TokenStream,
+{
+    let parse = (
+        opt(expr),
```

Comment:
In terms of semantics, this function should start parsing from `CASE` and the caller should not consume `CASE` before calling it.

```suggestion
pub fn expr_case<S>(input: &mut S) -> PResult<Expr>
where
    S: TokenStream,
{
    let parse = (
        Keyword::CASE,
        opt(expr),
```

Reply:
Unfortunately we have to following the v1's behavior.

Another problem is `dispatch!` is much more efficient than `alt`, even in v2, we have to use the style.

Reply:
The only thing I can do is rename it to `expr_case_suffix`.

Reply:
Okay let's correct this later.

`dispatch!` can also be used with `peek(keyword)` which don't consume the token. 😄